destination/writer/writer.go

```diff
@@ -0,0 +1,246 @@
+
+	if err = w.client.InsertRow(ctx, table, columns, values); err != nil {
+		return fmt.Errorf("insert row: %w", err)
+	}
+
+	return nil
+}
+
+// Close closes the firebolt connection.
+func (w *Writer) Close(ctx context.Context) error {
+	w.client.Close(ctx)
+
+	return nil
+}
+
+// getTableName returns either the records metadata value for table
+// or the default configured value for table.
+func (w *Writer) getTableName(metadata map[string]string) string {
+	tableName, ok := metadata[metadataTable]
+	if !ok {
+		return w.table
+	}
+
+	return strings.ToLower(tableName)
+}
+
+// structurizeData converts sdk.Data to sdk.StructuredData.
+func (w *Writer) structurizeData(data sdk.Data) (sdk.StructuredData, error) {
+	if data == nil || len(data.Bytes()) == 0 {
+		return nil, nil
+	}
+
+	structuredData := make(sdk.StructuredData)
+	if err := json.Unmarshal(data.Bytes(), &structuredData); err != nil {
+		return nil, fmt.Errorf("unmarshal data into structured data: %w", err)
+	}
+
+	// convert keys to lower case
+	structuredDataLower := make(sdk.StructuredData)
+	for key, value := range structuredData {
+		if parsedValue, ok := value.(map[string]any); ok {
+			jsonValue, err := json.Marshal(parsedValue)
+			if err != nil {
+				return nil, fmt.Errorf("marshal map into json: %w", err)
+			}
+
+			structuredDataLower[strings.ToLower(key)] = string(jsonValue)
+
+			continue
+		}
+
+		structuredDataLower[strings.ToLower(key)] = value
+	}
+
+	return structuredDataLower, nil
+}
+
+// extractColumnsAndValues turns the payload into slices of
+// columns and values for upserting into Firebolt.
+func (w *Writer) extractColumnsAndValues(payload sdk.StructuredData) ([]string, []any) {
+	var (
+		columns []string
+		values  []any
+	)
+
+	for key, value := range payload {
+		columns = append(columns, key)
+		values = append(values, value)
+	}
+
+	return columns, values
+}
+
+func (w *Writer) convertPayload(data sdk.StructuredData) (sdk.StructuredData, error) {
+	result := make(sdk.StructuredData, len(data))
+
+	for key, value := range data {
+		if value == nil {
+			result[key] = value
+
+			continue
+		}
+
+		switch reflect.TypeOf(value).Kind() {
+		case reflect.Map, reflect.Slice:
+			bs, err := json.Marshal(value)
+			if err != nil {
+				return nil, fmt.Errorf("marshal: %w", err)
+			}
+
+			result[key] = string(bs)
+
+			continue
+		}
+
+		switch w.columnTypes[strings.ToLower(key)] {
+		case typeDate:
+			v, ok := value.(time.Time)
+			if ok {
+				result[key] = v.Format("2006-01-02")
+
+				continue
+			}
+
+			valueStr, ok := value.(string)
+			if ok {
+				timeValue, err := parseToTime(valueStr)
+				if err != nil {
+					return nil, fmt.Errorf("convert value to time.Time: %w", err)
+				}
+
+				result[key] = timeValue.Format("2006-01-02")
+
+				continue
+			}
+
+			return nil, ErrInvalidTypeForDateColumn
+		case typeTimestamp:
+			v, ok := value.(time.Time)
+			if ok {
+				result[key] = v.Format("2006-01-02 15:04:05")
```

Comment:
Because firebolt timestamp support this format
https://docs.firebolt.io/general-reference/data-types.html#timestamp

I have added comment about it.